src/cryptonote_core/service_node_swarm.cpp

```diff
@@ -0,0 +1,273 @@
+  {
+    sorted_swarm_sizes.clear();
+    for (const auto &entry : swarm_to_snodes)
+    {
+      sorted_swarm_sizes.push_back({entry.first, entry.second.size()});
+    }
+    std::sort(sorted_swarm_sizes.begin(),
+              sorted_swarm_sizes.end(),
+              [](const swarm_size &a, const swarm_size &b) {
+                return a.size < b.size;
+              });
+  }
+
+  /// Assign each snode from snode_pubkeys into the FILL_SWARM_LOWER_PERCENTILE percentile of swarms
+  /// and run the excess/threshold logic after each assignment to ensure new swarms are generated when required.
+  prod_static void assign_snodes(const std::vector<crypto::public_key> &snode_pubkeys, swarm_snode_map_t &swarm_to_snodes, std::mt19937_64 &mt)
+  {
+    std::vector<swarm_size> sorted_swarm_sizes;
+    for (const auto &sn_pk : snode_pubkeys)
+    {
+      calc_swarm_sizes(swarm_to_snodes, sorted_swarm_sizes);
+      const size_t percentile_index = FILL_SWARM_LOWER_PERCENTILE * (sorted_swarm_sizes.size() - 1) / 100;
+      const size_t percentile_value = sorted_swarm_sizes.at(percentile_index).size;
+      /// Find last occurence of percentile_value
+      size_t upper_index = sorted_swarm_sizes.size() - 1;
+      for (size_t i = 0; i < sorted_swarm_sizes.size(); ++i)
+      {
+        if (sorted_swarm_sizes[i].size > percentile_value)
+        {
+          /// Would never happen for i == 0
+          upper_index = i - 1;
+          break;
+        }
+      }
+      const size_t random_idx = uniform_distribution_portable(mt, upper_index + 1);
+      const swarm_id_t swarm_id = sorted_swarm_sizes[random_idx].swarm_id;
+      swarm_to_snodes.at(swarm_id).push_back(sn_pk);
+      /// run the excess/threshold round after each additional snode
+      create_new_swarm_from_excess(swarm_to_snodes, mt);
+    }
+  }
+
+  /// This function needs to be called after each single snode is stolen
+  prod_static bool calc_robin_hood_round(const swarm_snode_map_t &swarm_to_snodes, std::vector<excess_pool_snode> &rich_snodes, std::vector<swarm_id_t> &poor_swarm_ids)
+  {
+    std::vector<swarm_size> sorted_swarm_sizes;
+    calc_swarm_sizes(swarm_to_snodes, sorted_swarm_sizes);
+    rich_snodes.clear();
+    poor_swarm_ids.clear();
+    const size_t rich_percentile_index = STEALING_SWARM_UPPER_PERCENTILE * (sorted_swarm_sizes.size() - 1) / 100;
+    size_t rich_percentile_value = sorted_swarm_sizes.at(rich_percentile_index).size;
+    /// MIN_SWARM_SIZE + 1 because we don't want to steal from snodes that would starve if we do so
+    rich_percentile_value = std::max(MIN_SWARM_SIZE, rich_percentile_value);
+    size_t deficit = 0;
+    size_t excess = 0;
+    for (const auto &entry : swarm_to_snodes)
+    {
+      const size_t swarm_size = entry.second.size();
+      if (swarm_size < MIN_SWARM_SIZE)
+      {
+        deficit += MIN_SWARM_SIZE - swarm_size;
+        poor_swarm_ids.push_back(entry.first);
+      }
+      else if (swarm_size > rich_percentile_value)
```

Comment:
Sacha and I discussed this over discord, here were my thoughts and the conclusion

You are selecting the 25th percentile swarm size (i.e. 75% of the swarms are greater or equal to this size), and then calculating the excess by counting the number of swarms *greater* than this value
I can imagine a situation where 76% of the swarms are all size 6 for example, this would result in the excess being calculated as zero and the algorithm would dissolve a swarm

Sacha said that if we changed the rich percentile to `rich_percentile_value = std::max(MIN_SWARM_SIZE + 1, rich_percentile_value);`
and then changed this check to be `if (swarm_size >= rich_percentile_value)` it would solve the problem

Reply:
Also discussed was the idea to change the logic to:
A) Completely fill starving swarms 1 at a time
B) Ensure we only attempt to fill a swarm if we have enough excess to do so

This is to stop the unnecessary stealing of a snode if we are going to be required to dissolve the swarm anyway

Reply:
Good summary. Amendments will come soon!